app/jobs/autoscaling.py

```diff
@@ -198,9 +208,133 @@ def set_desired_capacity(asg, compute_desired_instances, can_scale=True):
     if can_scale:
         if DEBUG:
             print cmd
+        if num_desired < previous_desired:
+            start_draining(asg, previous_desired - num_desired)
+        elif num_desired > previous_desired:
+            stop_draining(asg, num_desired - previous_desired)
         aws_command(cmd)
 
 
+def instances_in(asg):
+    return [item["InstanceId"] for item in asg["Instances"]]
+
+
+def iso_time_now():
+    return datetime.datetime.now().isoformat()
+
+
+def add_draining_tag(instance_ids):
+    cmd = "aws ec2 create-tags --resources {list_instances} --tags Key={draining_tag},Value={timestamp}"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), draining_tag=DRAINING_TAG, timestamp=iso_time_now())
+    aws_command(cmd)
+
+
+def remove_draining_tag(instance_ids):
+    delete_tags_from_instances(instance_ids, [DRAINING_TAG])
+
+
+def start_draining(asg, num_instances):
+    instance_ids = instances_to_drain(asg, num_instances)
+    add_draining_tag(instance_ids)
+
+
+def stop_draining(asg, num_instances):
+    instance_ids = instances_to_rescue(asg, num_instances)
+    add_termination_protection(instance_ids, asg)
+    remove_draining_tag(instance_ids)
+
+
+def instances_to_drain(asg, num_instances):
+    instance_ids = instances_in(asg)
+    cmd = "aws ec2 describe-instances --instance-ids {list_instances} --query 'Reservations[*].Instances[*].LaunchTime' --no-paginate --output text"
+    cmd = cmd.format(list_instances=' '.join(instance_ids))
+    launch_times = [dateutil.parser.parse(datestring) for datestring in aws_command(cmd).splitlines()]
+    instance_ids_sorted_by_increasing_launch_time = [item[0] for item in sorted(zip(instance_ids, launch_times), key=itemgetter(1))]
+    return instance_ids_sorted_by_increasing_launch_time[:num_instances]
+
+
+def instances_to_rescue(asg, num_instances):
+    '''
+    Rescue the instances that have been draining for the least amount of time.
+    That way we can probably terminate the non-rescued instances sooner.
+    '''
+    draining_instances = get_draining_servers(asg)
+    draining_instances_sorted = [key for key, value in sorted(draining_instances.items(), key=itemgetter(1), reverse=True)]
+    return draining_instances_sorted[:num_instances]
+
+
+def get_draining_servers(asg):
+    ''' returns a map of draining instance IDs to the time they started draining '''
+    cmd = "aws ec2 describe-tags --filters 'Name=tag-key,Values={draining_tag}' 'Name=resource-id,Values={instance_ids}'"
+    cmd = cmd.format(draining_tag=DRAINING_TAG, instance_ids=','.join(instances_in(asg)))
+    tag_dict = json.loads(aws_command(cmd))
+    instance_dict = { item['ResourceId']: dateutil.parser.parse(item['Value']) for item in tag_dict['Tags'] if item['Key'] == DRAINING_TAG }
+    return instance_dict
+
+
+def count_running_alignment_jobs(asg):
+    ''' returns a map of instance IDs to number of jobs running on the instance '''
```

Comment:
For this initial version, can we skip this altogether in favor of something much simpler, like checking the load average over last 5 minutes with `ssh <host> uptime | awk '{print $12}'`  to determine if that host is finished draining?

(Even thought checking tags was my suggestion at one point, I didn't intend it for this early stage, and I meant it more for enabling more complex protocols later.  I believe simplicity is something we should truly strive for, even to the point of removing optimizations.)

Reply:
Makes sense. But I really like the tags as a way of easily seeing what's happening in the gsnap/rapsearch servers via the console. I would vote to leave this feature in, since I've already tested it and it works and it can be useful later